azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -40,81 +41,243 @@
+            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
+            xml_doc = parse_doc(xml_text)
 
             role_instance = find(xml_doc, "RoleInstance")
             if role_instance:
                 break
             time.sleep(0.5)
         else:
-            raise IncompleteGoalStateError("Fetched goal state without a RoleInstance [incarnation {inc}]".format(inc=self.incarnation))
+            incarnation = findtext(xml_doc, "Incarnation")
+            raise ProtocolError("Fetched goal state without a RoleInstance [incarnation {inc}]".format(inc=incarnation))
 
-        try:
-            self.role_instance_id = findtext(role_instance, "InstanceId")
-            role_config = find(role_instance, "Configuration")
-            self.role_config_name = findtext(role_config, "ConfigName")
-            container = find(xml_doc, "Container")
-            self.container_id = findtext(container, "ContainerId")
-
-            AgentGlobals.update_container_id(self.container_id)
-
-            # these properties are populated by fetch_full_goal_state()
-            self._hosting_env_uri = findtext(xml_doc, "HostingEnvironmentConfig")
-            self.hosting_env = None
-            self._shared_conf_uri = findtext(xml_doc, "SharedConfig")
-            self.shared_conf = None
-            self._certs_uri = findtext(xml_doc, "Certificates")
-            self.certs = None
-            self._remote_access_uri = findtext(container, "RemoteAccessInfo")
-            self.remote_access = None
-            # TODO: extensions_config is an instance member only temporarily. Once we stop comparing extensionsConfig with
-            # vmSettings, it will be replaced with the extensions goal state
-            self.extensions_config = None
-            self._extensions_config_uri = findtext(xml_doc, "ExtensionsConfig")
+        # Telemetry and the HostGAPlugin depend on the container id/role config; keep them up-to-date each time we fetch the goal state
+        # (note that these elements can change even if the incarnation of the goal state does not change)
+        container = find(xml_doc, "Container")
+        container_id = findtext(container, "ContainerId")
+        role_config = find(role_instance, "Configuration")
+        role_config_name = findtext(role_config, "ConfigName")
 
-        except Exception as exception:
-            # We don't log the error here since fetching the goal state is done every few seconds
-            raise ProtocolError(msg="Error fetching goal state", inner=exception)
+        AgentGlobals.update_container_id(container_id)  # Telemetry uses this global to pick up the container id
 
-    def fetch_full_goal_state(self, wire_client):
-        try:
-            logger.info('Fetching goal state [incarnation {0}]', self.incarnation)
+        wire_client.update_host_plugin(container_id, role_config_name)
+
+        return xml_text, xml_doc
+
+    def _fetch_vm_settings(self, force_update=False):
```

Comment:
The extensions goal state state is now a property of GoalState so we fetch it here